### ISSUE TITLE
Reviewer feedback: Better motivate why demonstrating a ReCOGS/COGS solution with a non-tree structured / non-hierarchical representation is useful/interesting

### ISSUE THREAD

**Issue by willy-b** [willy-b/RASP-for-ReCOGS#70]
One of our excellent reviewers brought up that we do not sufficiently motivate why a non-tree structured/non-hierarchical flat pattern matching solution to semantic parsing for ReCOGS/COGS is interesting and they note we should mention e.g. the examples cited Yao & Koller, EMNLP 2022 of models with a tree-structured or hierarchical design that perform COGS perfectly. (We will add such content as part of this issue.)

This author also notes that in the conclusion of the original COGS paper (Kim and Linzen 2020), the COGS authors suggest architecturally tree-structured/hierarchical models might be required to solve the structural generalizations:
“What architecture would be needed to solve COGS? For structural generalization cases, the results of Bowman et al. (2015); Evans et al. (2018) and McCoy et al. (2019) suggest that tree- structured models may provide a better inductive bias. In particular, Bowman et al. (2015) showed that tree-structured neural networks generalized to longer sequences.”

Separate from adding new references as requested for this issue, below is a restatement of our original argument for the interest in such an approach using existing references:

**Why proving the Transformer can solve ReCOGS/COGS with a flat/non-hierarchical solution at all is interesting separate from claims on learnability or required training data**

We cite and quote (Csordás et al., 2022) in the paper, who state what we might call Csordás’ curse, that "[for compositional operations] the network should be sufficiently deep, at least as deep as the deepest data dependency in the computational graph built from elementary operations (e.g., in the case of a parse tree, this is the depth of the tree)". Though (Csordás et al., 2022) (and also Universal Transformers (Dehghani et al., 2019)) suggest reusing the same weights for each layer to avoid this (and other modified “tree-structured” architectures force depth symmetry in other ways), in a vanilla Transformer, this would also mean the same compositional operations would need to be relearned by the weights at each layer depth independently, providing one clear mechanism for a failure to generalize to new sequence lengths if the new lengths came from unseen parse tree depths (e.g. as in the COGS/ReCOGS prepositional phrase recursion generalization splits, where training on depths 0-2 is then tested on depths up to recursion depth 12 (see ReCOGS or COGS paper)).

Such recursion depth generalizations are not a problem in a flat/non-hierarchical approach to ReCOGS like we prove a vanilla Transformer is capable of in this work, which has the computational advantage of a constant layer depth requirement regardless of recursion depth (using Decoder loop to unroll the (tail) recursive aspects in COGS/ReCOGS, with each Encoder/Decoder pass for each token prediction requiring constant depth independent of sequence length, and thus also no untrained layers that would require deeper data dependencies to train as in Csordas' expectation). Note we also link to a working RASP-for-COGS (NOT just ReCOGS) implementation in our work (available in the public repo for the identifiable paper or associated 4open.science repo version for the deidentified review version of the paper at time of initial submission).

(In fact based on the Wu et al 2023 baseline Encoder-Decoder Transformer failure on v_dat_p2 recipient pp modification generalization (Figure 3; not predicted to be difficult if using a hierarchical representation) and commission of "attraction" errors (Figure 7; not predicted to our knowlege to occur without flat/linear pattern matching), I think we have some evidence the Wu et al 2023 baseline Encoder-Decoder Transformer is also using a similar flat-pattern matching strategy but without (unfortunately) having learned to mask out / ignore the “pp det common_noun” and “pp proper_noun” attractors everywhere. 
Though I am NOT claiming that very deep Transformer networks used in practice for language modeling are using flat pattern matching, and there is plenty of evidence that they do use hierarchical representations (e.g. “A structural probe for finding syntax in word representations” (Hewitt and Manning 2019) cited in the paper).)

**Comment by willy-b** [comment]
Added to above issue description:
```
Note, one might argue that such a flat pattern matching approach has somehow cheated or side-stepped the problem of learning compositional rules since it doesn't look like context-free grammar parsing anymore (which is what generated the input sentences). However, for example considering the masking of pp noun trick followed by flat pattern matching the main grammar pattern (verb type, "to"/"by", nouns), is it really cheating to avoid hierarchically combining the "noun1 beside the noun2 on the noun3" in multiple steps into a single noun phrase? The output logical form required by the task completely separates the prepositional phrase information (nmods) from the verb relationships (grammar pattern of sentence excluding the prepositional phrases), so there is no need for this information to be mixed together during the parsing and nothing is lost by just looping over the preposition relationships at the end in the decoder loop. The operation of associating each noun via a preposition with the other noun it modifies is preserved, as are the operations that relate verbs the sentences to the nouns they interact with, the processing of these two things has just been separated out by masking.
```

**Comment by willy-b** [comment]
Re:
> In fact based on the Wu et al 2023 baseline Encoder-Decoder Transformer failure on v_dat_p2 recipient pp modification 
> generalization (Figure 3; not predicted to be difficult if using a hierarchical representation) and commission of "attraction" 
> errors (Figure 7; not predicted to our knowlege to occur without flat/linear pattern matching), I think we have some 
> evidence the Wu et al 2023 baseline Encoder-Decoder Transformer is also using a similar flat-pattern matching strategy but 
> without (unfortunately) having learned to mask out / ignore the “pp det common_noun” and “pp proper_noun” attractors everywhere.

See also https://aclanthology.org/2024.naacl-long.402/ "The Impact of Depth on Compositional Generalization in Transformer Language Models" by Petty et al already discussed multiple places in our paper (they find increasing the depth to 32 layers does not improve structural generalization on COGS by Transformers which is consistent with our explanation of the Wu et al 2023 baseline model using a non-hierarchical strategy for the recursion splits).

**Comment by willy-b** [comment]
Made some small changes towards this in #75 into #12
